File: Toolbox/nt_get_events.m
function events = nt_get_events(measures,params)
%nt_get_events Create event table from NoviTrack markers.
%
%  EVENTS = nt_get_events(MEASURES,PARAMS)
%
%  Events are derived from measures.markers and should not be stored in
%  measures. Keeping only markers in saved databases avoids storing MATLAB
%  table objects, which do not round-trip cleanly through Python.
%
% 2026, Alexander Heimel

if nargin<2 || isempty(params)
    params = nt_load_parameters();
end

if isempty(measures) || ~isfield(measures,'markers') || isempty(measures.markers)
    events = table([],strings(0,1),'VariableNames',{'time','event'});
    return
end

events = table([measures.markers.time]',...
    string({measures.markers.marker}'),'VariableNames',{'time','event'});

% temporary fix for illegal field names 0 and 1
events.event(events.event == "0") = "opto_off";
events.event(events.event == "1") = "opto_on";

if params.use_clean_baseline
    i = 1;
    % remove all same events within pretime
    while i<height(events)
        events(events.time>events.time(i) & events.time<(events.time(i) + params.nt_pretime) & events.event == events.event(i),:) = [];
        i = i + 1;
    end
end

if params.use_ultraclean_baseline
    i = 1;
    % remove all events within pretime
    while i<height(events)
        events(events.time>events.time(i) & events.time<(events.time(i) + params.nt_pretime),:) = [];
        i = i + 1;
    end
end

end

File: novitrack/nt_get_events.py
"""Create NoviTrack event tables from marker annotations."""

from __future__ import annotations

from collections.abc import Mapping
from typing import Any

import pandas as pd


def _get(obj: Any, name: str, default: Any = None) -> Any:
    if obj is None:
        return default
    if isinstance(obj, Mapping):
        return obj.get(name, default)
    return getattr(obj, name, default)


def nt_get_events(measures: Any, params: Any | None = None) -> pd.DataFrame:
    """Create an events DataFrame from ``measures["markers"]``.

    This mirrors ``nt_get_events.m``. Events are derived on demand so saved
    databases do not need to store MATLAB table objects.
    """
    markers = _get(measures, "markers", None)
    if markers is None or len(markers) == 0:
        return pd.DataFrame({"time": pd.Series(dtype=float), "event": pd.Series(dtype=str)})

    events = pd.DataFrame(
        {
            "time": [float(_get(marker, "time")) for marker in markers],
            "event": [str(_get(marker, "marker")) for marker in markers],
        }
    )
    events["event"] = events["event"].replace({"0": "opto_off", "1": "opto_on"})

    pretime = float(_get(params, "nt_pretime", 10))

    if bool(_get(params, "use_clean_baseline", False)):
        index = 0
        while index < len(events):
            row = events.iloc[index]
            remove = (
                (events["time"] > row["time"])
                & (events["time"] < row["time"] + pretime)
                & (events["event"] == row["event"])
            )
            events = events.loc[~remove].reset_index(drop=True)
            index += 1

    if bool(_get(params, "use_ultraclean_baseline", False)):
        index = 0
        while index < len(events):
            row = events.iloc[index]
            remove = (events["time"] > row["time"]) & (events["time"] < row["time"] + pretime)
            events = events.loc[~remove].reset_index(drop=True)
            index += 1

    return events
